Enhancement - Notification timeout and addition of Notification page

microtechno9000 opened the pull request:
# Description
Enhancement to ARM in how notifications are handled. As suggested in #790 

1. Notification pop-up (bootstrap toast) timeout is now a user configurable item within UISettings
2. Users can modify the Notification Timeout from the UI Settings, with a default loaded in with the database update of 6.5 seconds
3. Addition of a notification bell on the ARM Menu bar
4. Notificaiton bell shows the number of current 'unseen' / 'not closed' notifications as a number (see image)
5. Clicking the notification bell brings up the current list of 'unseen' / 'not closed' notifications (see image)

Fixes #790

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update
- once reviewed, will update the Wiki

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce.
Please also list any relevant details for your test configuration

- [x] Docker
- [x] Ubuntu 20.04

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have tested that my fix is effective or that my feature works

# Changelog:
as in description

# Logs
No change to the ARM ripper, no log attached.
![ARM_notification_01](https://user-images.githubusercontent.com/62650032/228112403-2e51babc-c3ce-4165-8abb-6f98f5e49f11.png)
![ARM_notification_02](https://user-images.githubusercontent.com/62650032/228112421-2aef280e-9c4a-4814-8012-adf9afd3a6e1.png)
![ARM_notification_03](https://user-images.githubusercontent.com/62650032/228112652-caa4b04c-293c-4b17-a3a1-e440811a07a1.png)

microtechno9000 commented:
security hot spot is linking to the /login page, I don't see any security issue with the way this request is called/run. Not sure how to close the code assesemnt

1337-server commented:
After migrating the database, I run into this error on the database page.

`'NoneType' object has no attribute 'database_limit'`

This was before running another disc post upgrade. Inserting a disc seems to resolve this, however another error triggered and killed the UI and the currently running job.
```
DEBUG ARM: json_api.get_x_jobs jobs  - we have 1 jobs
Traceback (most recent call last):
File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2091, in __call__
return self.wsgi_app(environ, start_response)
File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2076, in wsgi_app
     response = self.handle_exception(e)
   File "/usr/local/lib/python3.8/dist-packages/flask_cors/extension.py", line 165, in wrapped_function
     return cors_after_request(app.make_response(f(*args, **kwargs)))
   File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2073, in wsgi_app
    response = self.full_dispatch_request()
   File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1518, in full_dispatch_request
     rv = self.handle_user_exception(e)
   File "/usr/local/lib/python3.8/dist-packages/flask_cors/extension.py", line 165, in wrapped_function
     return cors_after_request(app.make_response(f(*args, **kwargs)))
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1516, in full_dispatch_request
    rv = self.dispatch_request()
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1502, in dispatch_request
     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
   File "/usr/local/lib/python3.8/dist-packages/flask_login/utils.py", line 272, in decorated_view
     return func(*args, **kwargs)
   File "/opt/arm/arm/ui/jobs/jobs.py", line 239, in feed_json
     return_json = valid_modes[mode]['funct'](*args)
TypeError: get_notify_timeout() takes 0 positional arguments but 1 was given
 Traceback (most recent call last):
   File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2091, in __call__
    return self.wsgi_app(environ, start_response)
  File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2076, in wsgi_app
     response = self.handle_exception(e)
  File "/usr/local/lib/python3.8/dist-packages/flask_cors/extension.py", line 165, in wrapped_function
     return cors_after_request(app.make_response(f(*args, **kwargs)))
   File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 2073, in wsgi_app
     response = self.full_dispatch_request()
   File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1518, in full_dispatch_request
     rv = self.handle_user_exception(e)
   File "/usr/local/lib/python3.8/dist-packages/flask_cors/extension.py", line 165, in wrapped_function
     return cors_after_request(app.make_response(f(*args, **kwargs)))
   File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1516, in full_dispatch_request
     rv = self.dispatch_request()
   File "/usr/local/lib/python3.8/dist-packages/flask/app.py", line 1502, in dispatch_request
     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)
   File "/usr/local/lib/python3.8/dist-packages/flask_login/utils.py", line 272, in decorated_view
     return func(*args, **kwargs)
   File "/opt/arm/arm/ui/jobs/jobs.py", line 239, in feed_json
     return_json = valid_modes[mode]['funct'](*args)
 TypeError: get_notify_timeout() takes 0 positional arguments but 1 was given 
```

The UI restarted itself, but the job was zombied.

It also seems like the notifications were dismissed without any input.

microtechno9000 commented:
will have a look, seems the JS is complaining.

At the moment, the notification will timeout based on the toast timeout and then disappear. I haven't added in any new logic to assess if it timeout out or has been seen.

if preferred, I could modify the timeout to not mark as seen, then keep the notification till cleared on the notification page

microtechno9000 commented:
@1337-server resolved the issue with the notification. Also changed the behavior
- New notifications will be displayed for the UIDisplay time, then fade out and not show again
- Added a new field in for the notification pane, the timed out notifications will still be shown but only on this page, using a seperate value 'cleared' and 'cleared_time' in the database

microtechno9000 commented:
Don't merge yet, something is broken with ripping disks. need to check it first

shitwolfymakes commented:
Please fix the bug described [here](https://sonarcloud.io/project/issues?id=automatic-ripping-machine_automatic-ripping-machine&pullRequest=807&resolved=false&types=BUG)

microtechno9000 commented:
> Please fix the bug described [here](https://sonarcloud.io/project/issues?id=automatic-ripping-machine_automatic-ripping-machine&pullRequest=807&resolved=false&types=BUG)

@shitwolfymakes fixed the bug

also in checking the bug above and testing the notifications work with a drive (new job), I noticed a bug in the new drive search. Resolved and added to the latest commit

shitwolfymakes commented on arm/ui/database/database.py:
```diff
@@ -36,6 +36,8 @@ def view_database():
     Outputs every job from the database
      this can cause serious slow-downs with + 3/4000 entries
     """
+    global armui_cfg
```
Not super fond of using a global for this, but it's not an issue

microtechno9000 replied:
I would prefer to attach it to the flask app as a config of sorts to that, but not sure how that would work. Plus would change a few pages/functions in ARM.
Unfortunately without the global I couldn't get this function to work correctly without throwing errors.

shitwolfymakes commented on arm/ui/settings/DriveUtils.py:
```diff
@@ -27,7 +27,7 @@ def drives_search():
 
     for device in context.list_devices(subsystem='block'):
-        regexoutput = re.search(r'(/dev/sr\d)', device.device_node)
+        regexoutput = re.search(r'(\/dev\/sr\d{1,2})', device.device_node)
```
This will allow for up too 100 drives, so I don't imagine we'll need to change it to a number of any length

microtechno9000 replied:
the current code will only allow for sr0 through sr9 and won't detect anything above that, as you noted the update will resolve sr0 through sr99. More than enough. I would like to see a home PC with 99 drives

shitwolfymakes commented on arm/ui/utils.py:
```diff
@@ -201,12 +201,26 @@ def arm_db_cfg():
     else:
-        armui_cfg = models.UISettings.query.get(1)
-        app.jinja_env.globals.update(armui_cfg=armui_cfg)
+        # if the database has been updated
```
Is this a duplicate of the above?

microtechno9000 replied:
yes, it is a duplicate. I have removed the else statement and duplicate code.
Nice pickup

shitwolfymakes reviewed (changes requested):
Looks good, but I have question about ui/utils.py

microtechno9000 commented:
Suggest holding off on this PR till #819 is completed, to confirm no issues with the database updates

shitwolfymakes commented:
Copy, sounds good

shitwolfymakes commented on arm/ui/json_api.py:
```diff
@@ -445,3 +445,20 @@ def read_notification(notify_id):
     return return_json
+
+
+def get_notify_timeout(notify_timeout):
```
```suggestion
def get_notify_timeout():
```

shitwolfymakes reviewed (changes requested):
Fix this real quick

microtechno9000 commented:
> Fix this real quick

Not fixed, as per the comment against the review in the code. The function in jobs.py needs an id, otherwise an error is thrown

https://github.com/automatic-ripping-machine/automatic-ripping-machine/blob/b4c6b69bb669908aa76650ed9c7b8bbe7eb7d140/arm/ui/jobs/jobs.py#L235
